tools/clang/include/clang/SPIRV/Structure.h

```diff
@@ -162,50 +168,87 @@ class SPIRVModule {
   };
 
   /// \brief The struct representing an entry point.
   struct EntryPoint {
     inline EntryPoint(spv::ExecutionModel, uint32_t id, std::string name,
                       std::initializer_list<uint32_t> interface);
 
-    spv::ExecutionModel executionModel;
-    uint32_t targetId;
-    std::string targetName;
-    std::initializer_list<uint32_t> interfaces;
+    const spv::ExecutionModel executionModel;
+    const uint32_t targetId;
+    const std::string targetName;
+    const std::initializer_list<uint32_t> interfaces;
   };
 
   /// \brief The struct representing a debug name.
   struct DebugName {
     inline DebugName(uint32_t id, llvm::Optional<uint32_t> index,
                      std::string targetName);
 
-    uint32_t targetId;
-    llvm::Optional<uint32_t> memberIndex;
-    std::string name;
+    const uint32_t targetId;
+    const llvm::Optional<uint32_t> memberIndex;
+    const std::string name;
+  };
+
+  /// \brief The struct representing a deocoration and its target <result-id>.
+  struct DecorationIdPair {
+    inline DecorationIdPair(const Decoration &decor, uint32_t id);
+
+    const Decoration &decoration;
+    const uint32_t targetId;
+  };
+
+  /// \brief The struct representing a type and its <result-id>.
+  struct TypeIdPair {
+    inline TypeIdPair(const Type &ty, uint32_t id);
+
+    const Type &type;
+    const uint32_t resultId;
+  };
+
+  /// \brief The struct representing a constant and its type.
+  struct Constant {
+    inline Constant(const Type &ty, Instruction &&value);
+    const Type &type;
+    Instruction constant;
   };
 
   Header header; ///< SPIR-V module header.
   std::vector<spv::Capability> capabilities;
   std::vector<std::string> extensions;
-  std::vector<std::pair<uint32_t, std::string>> extInstSets;
+  std::vector<ExtInstSet> extInstSets;
+  // addressing and memory model must exist for a valid SPIR-V module.
+  // We make them optional here just to provide extra flexibility of
+  // the representation.
   llvm::Optional<spv::AddressingModel> addressingModel;
   llvm::Optional<spv::MemoryModel> memoryModel;
   std::vector<EntryPoint> entryPoints;
-  // XXX: Right now the following are basically vectors of Instructions.
-  // They will be turned into vectors of more full-fledged classes gradually
-  // as we implement more features.
   std::vector<Instruction> executionModes;
-  // TODO: support other debug instructions
+  // TODO: source code debug information
   std::vector<DebugName> debugNames;
-  std::vector<Instruction> decorations;
-  std::vector<Instruction> typesValues;
-  std::vector<Function> functions;
+  std::vector<DecorationIdPair> decorations;
+  // Note that types and constants are interdependent; Types like arrays have
+  // <result-id>s for constants in their definition, and constants all have
+  // their corresponding types. We store types and constants separately, but
+  // they should be handled together.
+  llvm::MapVector<const Type *, uint32_t> types;
```

Comment:
Yes. We need the order preserved b/c types can depend on each other. E.g., for `float4`, we need to define the `OpFloat` first and the the `OpVector`.